### DISCUSSION TITLE
EPG not working in "reverse" mode

### DISCUSSION THREAD

**Discussion by adamsdavis572** [euzu/tuliprox#27] [General]
@euzu i have the xtream reverse proxy mode configured and working now. This is a fantastic addition - thanks very much. 

However i dont seem to be receiving EPG in this mode

The call to m3u-filter proxy happens and it returns a 200 

`[2023-12-06T18:27:08Z INFO  actix_web::middleware::logger] 192.168.1.73 "GET /xmltv.php?username=<username>&password=<password> HTTP/1.1" 200 749721 "-" "Dalvik/2.1.0 (Linux; U; Android 14; SM-S906B Build/UP1A.231005.007)" 0.003312`

When i call this URL in a browser it does return the xml epg list but i cant see anything in Smarters 

Any ideas?

**Comment by euzu** [comment]
@adamsdavis572 i have tested the epg, and it is working. 

If you have given an epg_url -> the epg is downloaded during update and written and served from/to disc (the advantage is, the epg is reduced to the filtered channels, unecessary epg is removed).

If you have used epg_url and then removed it, the old file will remain on disc.

The xtream epg api will first look for the file on disc, if it exists it is served, if not  the original povider url is accessed and the content is passed to the client (nothing changed as is from provider).

I have tested it with android smarters light, and it is working as expected.

How it works: the channels have an attribute "epg_channel_id", this has to match the channel id inside the epg xml file.


What you can do: download the epg file with curl and check the channel id against the `epg_channel_id` in  `<m3u-filter-folder>/data/<target-name>/live.json` . If they match , the epg should be available

**Reply by adamsdavis572** [reply]
@euzu i have removed any reference to epg from the config.xml and deleted the files in /data and restarted the service.  

```
sources:
- inputs:
    - type: xtream
      persist: playlist_1_{}
      url: 'http://<url>:8080'
      username: <username>
      password: <password>
      options: {xtream_info_cache: true, }
``` 
if i look at "bbc1.uk" for example (present in `<m3u-filter-folder>/data/<target-name>/live.json`),  there  are only 2 entries returned for bbc1.uk for `http://<m3u-filyer-ipaddress>:8901/xmltv.php?username=<username>&password=<password>`

```
    <channel id="bbc1.uk">
        <display-name>UK| BBC 1 HD</display-name>
    </channel>
```

```
    <programme channel="bbc1.uk" start="20231206091500 +0000" stop="20231206100000 +0000">
        <title>Morning Live at Christmas</title>
        <desc>Conversation, advice and great company to make your morning. [S] [HD.</desc>
    </programme>
```

The only programme data is for yesterday and its a single programme. 

have i missed something?

**Reply by adamsdavis572** [reply]
Also if i manually add the iptv provider epg to smarters as an alternate epg it does work fine.

**Comment by adamsdavis572** [comment]
once again apologies - i thought i had deleted all the persisted epg files in data but i had not!! i have removed it and epg is working now :-) ill play around with config for caching locally. This is working really well. the filtered and locally cached xtream files is much quicker loading than my previous solution. much appreciated
